resin/resin/imgui/sdf_tree.cpp

```diff
@@ -2,140 +2,322 @@
+  auto source_id = get_curr_payload();
 
-  is_selected = selected_ == node.node_id() || is_parent_selected_;
+  bool is_node_selected = is_parent_selected_ || selected_ == node.node_id();
+  bool is_node_dragged  = is_parent_dragged_ || (source_id.has_value() && *source_id == node.node_id());
 
   ImGui::PushID(static_cast<int>(node.node_id().raw()));
-  if (ImGui::Selectable(node.name().data(), &is_selected)) {
+  if (is_node_dragged) {
+    ImGui::BeginDisabled();
+  }
+
+  ImGuiTreeNodeFlags flags = ImGuiTreeNodeFlags_Leaf | ImGuiTreeNodeFlags_NoTreePushOnOpen |
+                             ImGuiTreeNodeFlags_SpanAvailWidth | ImGuiTreeNodeFlags_FramePadding;
+
+  if (is_node_selected) {
+    flags |= ImGuiTreeNodeFlags_Selected;
+  }
+
+  // TODO(SDF-100): Use primitive icons
+  ImGui::PushStyleVar(ImGuiStyleVar_FramePadding, ImVec2(0.F, 2.F));
+  ImGui::TreeNodeEx(node.name().data(), flags);
+  ImGui::PopStyleVar();
+
+  if (ImGui::IsItemClicked()) {
     selected_ = node.node_id();
   }
+
+  if (is_node_dragged) {
+    ImGui::EndDisabled();
+  }
+
+  if (!is_node_dragged) {
+    drag_and_drop(node, true);
+  }
+
+  render_op(node);
+
   ImGui::PopID();
 }
 
-void SDFTreeOperationVisitor::visit_group(::resin::GroupNode& node) {
-  if (op == SDFTreeOperationVisitor::Operation::PushGroup) {
-    node.push_back_child<::resin::GroupNode>(::resin::SDFBinaryOperation::Union)
-        .push_back_child<::resin::SphereNode>(::resin::SDFBinaryOperation::Union);
-  } else if (op == SDFTreeOperationVisitor::Operation::PushPrimitive) {
-    node.push_back_child<::resin::SphereNode>(::resin::SDFBinaryOperation::Union);
+void SDFTreeComponentVisitor::render_op(::resin::SDFTreeNode& node) const {
+  // TODO(SDF-100): Use operation icons
+  ImGui::SameLine(ImGui::GetWindowWidth() - 42.F);
+  if (is_first_ && node.bin_op() != ::resin::SDFBinaryOperation::Union) {
+    ImGui::TextColored(ImGui::GetStyleColorVec4(ImGuiCol_TextDisabled), " (%s)",
+                       kOperationSymbol.get_value(node.bin_op()).data());
+    if (ImGui::BeginItemTooltip()) {
+      ImGui::PushTextWrapPos(ImGui::GetFontSize() * 36.0F);
+      ImGui::TextUnformatted("This operation is ignored for the first element in the group");
+      ImGui::PopTextWrapPos();
+      ImGui::EndTooltip();
+    }
+  } else {
+    ImGui::Text(" (%s)", kOperationSymbol.get_value(node.bin_op()).data());
   }
 }
 
-void SDFTreeOperationVisitor::visit_primitive(::resin::PrimitiveNode& node) {
-  if (op == SDFTreeOperationVisitor::Operation::PushGroup) {
-    node.parent()
-        .push_back_child<::resin::GroupNode>(::resin::SDFBinaryOperation::Union)
-        .push_back_child<::resin::SphereNode>(::resin::SDFBinaryOperation::Union);
-  } else if (op == SDFTreeOperationVisitor::Operation::PushPrimitive) {
-    node.parent().push_back_child<::resin::SphereNode>(::resin::SDFBinaryOperation::Union);
+void SDFTreeComponentVisitor::render_tree(::resin::SDFTree& tree) {
```

Comment:
Some comments on behavior:
- [ ] minimal width and height
- [ ] some way to remove selections, i.e. by clicking the background of node list
- [ ] there seems to be some problem regarding padding (fig1)

Fig1:
![image](https://github.com/user-attachments/assets/6bffd807-65da-402c-8ca5-fbbda45d8365)

Reply:
> minimal width and height
Sadly I couldn't find any way to define the constraints for components and enforce them on the windows. For that reason we should constraint the windows instead:(